File: crates/agent/tests/frontmatter.rs
//! A KEY NOTHING READS IS A SETTING THAT LOOKS APPLIED.
//!
//! `spec::yaml` refuses every VALUE it cannot honour — `engine: reakt`,
//! `compact_at: lots`, a `tools:` line that is not a list — and until this
//! increment it silently dropped every KEY it did not know. That is the same
//! bug one level up: `temprature: 0.7` parses clean, changes nothing, and stays
//! in the file being believed. A frontmatter key is a small closed vocabulary,
//! so an unknown one is a mistake and never an extension.
//!
//! The shipped files are read here too, because a refusal this strict is only
//! safe if what the app actually deploys goes through it.

use agent::{parse_agent_file, AgentError};

const MAIN: &str = include_str!("../../../public/agents/main/agent.md");

fn refusal(file: &str) -> String {
    match parse_agent_file("fixture", file) {
        Err(AgentError::MalformedAgentFile { message, .. }) => message,
        Err(other) => panic!("an agent file refuses as its own typed error: {other:?}"),
        Ok(_) => panic!("that file should not have parsed"),
    }
}

/// The key is named, and so are the keys that exist — a refusal a person can
/// act on without opening the source.
#[test]
fn an_unknown_frontmatter_key_is_refused_and_says_what_the_keys_are() {
    let message = refusal("---\nname: helper\ntemprature: 0.7\n---\n\nBody.\n");
    assert!(message.contains("temprature"), "{message}");
    assert!(message.contains("temperature"), "the real key is offered: {message}");
    // A near-miss on a LIST key is the dangerous one: `stage:` would have left
    // the loop empty and the agent silently stage-less.
    let message = refusal("---\nname: helper\nstage: [plan]\n---\n\nBody.\n");
    assert!(message.contains("stage"), "{message}");
}

/// …and the two things that are NOT keys still are not. A YAML comment and a
/// blank line are skipped before any of this, which is what lets the shipped
/// `main` explain itself in its own frontmatter.
#[test]
fn comments_and_blank_lines_are_not_keys() {
    let file = "---\nname: helper\n\n# stages: [plan] — why this file does not\n---\n\nBody.\n";
    let spec = parse_agent_file("fixture", file).expect("a comment is not a setting");
    assert_eq!(spec.name, "helper");
    assert!(spec.stages.is_empty(), "a commented-out key is not applied");
}

/// THE SHIPPED FILE GOES THROUGH IT. A parser this strict is only trustworthy
/// if the app's own agents pass — a refusal nobody runs against the real files
/// is a deploy that comes up with no agents at all.
#[test]
fn the_shipped_agent_file_carries_only_keys_that_are_read() {
    parse_agent_file("main", MAIN).expect("the shipped main agent parses");
}

/// TWO LISTS OF THE SAME VOCABULARY, HELD IN AGREEMENT BY A TEST.
///
/// `yaml::KEYS` is what the refusal above prints; the `match` in
/// `yaml::set_field` is what actually accepts a key. Nothing in the compiler
/// keeps them the same, and `agent::faculty` documents what this codebase paid
/// for the last time a `match` and a separate `const` drifted: a new entry got
/// ZERO structural coverage while every gate stayed green. So every name the
/// refusal offers is walked through the reader here, and a name outside both is
/// refused — a key added to one and not the other fails HERE rather than
/// shipping as a setting that looks applied.
///
/// The values are the least each key will accept: this asserts the KEY is
/// known, and the value rules have their own tests.
#[test]
fn every_key_the_refusal_offers_is_a_key_the_reader_accepts() {
    let offered = refusal("---\nname: helper\nnosuchkey: x\n---\nBody.\n");
    let names: Vec<String> = offered
        .split("the keys are: ")
        .nth(1)
        .expect("the refusal prints the vocabulary")
        .split(", ")
        .map(|k| k.trim().to_string())
        .collect();
    assert!(names.len() >= 17, "every key is offered, goal.* included: {names:?}");
    for key in &names {
        let value = match key.as_str() {
            "stages" => "[work]",
            "tools" | "faculties" => "[]",
            "compact_at" | "keep_recent" | "max_rounds" | "passes" => "1",
            "temperature" => "0.1",
            "engine" => "react",
            "role" => "entry",
            // A goal needs its other halves to survive `refuse_contradictions`;
            // this test is about the READER knowing the key, so the shape it
            // is given is a whole legal goal.
            k if k.starts_with("goal.") => "x",
            _ => "x",
        };
        let goal = match key.starts_with("goal.") {
            true => "space: research\nstages: [work]\ngoal.outcome: o\ngoal.check: c\n",
            false => "",
        };
        let file = format!("---\nname: helper\n{goal}{key}: {value}\n---\nBody.\n");
        parse_agent_file("fixture", &file)
            .unwrap_or_else(|e| panic!("'{key}' is offered by the refusal but not read: {e:?}"));
    }
    // …and a name in neither place is still refused.
    assert!(parse_agent_file("fixture", "---\nname: h\ngoal.chck: x\n---\nB.\n").is_err());
}

/// THE GLOSS IS PART OF THE PARSER'S VOCABULARY, AND NOW SOMETHING SAYS SO.
///
/// `ui/authoring/key_help.rs` claims in its own doc comment that "a key that
/// exists in the parser and not here is a visible gap rather than a silent
/// one". That was true only of a reader who happened to compare the two lists:
/// nothing enforced it, and five keys — `faculties`, `passes` and the three
/// dotted `goal.*` — had already gone missing, while `role` was still glossed
/// with `summarizer`, a job deleted from `spec::ROLES`. The panel whose whole
/// subject is what an agent file may say was the thing saying it wrongly (I16).
///
/// The source is read as TEXT because the gloss lives in the UI crate, which
/// this pure one cannot depend on (I3). The refusal is the same machine-read
/// vocabulary `every_key_the_refusal_offers_is_a_key_the_reader_accepts` uses,
/// so neither side of this comparison is typed out here.
const KEY_HELP: &str = include_str!("../../ui/src/authoring/key_help.rs");

fn glossed() -> Vec<(String, String)> {
    let block = KEY_HELP
        .split("const KEYS: &[(&str, &str)] = &[")
        .nth(1)
        .expect("the gloss list is where this test says it is")
        .split("\n];")
        .next()
        .expect("…and it ends");
    block
        .lines()
        .filter_map(|l| l.trim().strip_prefix("(\""))
        .filter_map(|l| l.split_once("\", \""))
        .map(|(key, said)| (key.to_string(), said.trim_end_matches("\"),").to_string()))
        .collect()
}

#[test]
fn every_key_the_parser_reads_is_glossed_for_the_person_writing_the_file() {
    let offered = refusal("---\nname: helper\nnosuchkey: x\n---\nBody.\n");
    let mut parser: Vec<String> = offered
        .split("the keys are: ")
        .nth(1)
        .expect("the refusal prints the vocabulary")
        .split(", ")
        .map(|k| k.trim().to_string())
        .collect();
    let mut shown: Vec<String> = glossed().into_iter().map(|(k, _)| k).collect();
    assert!(!shown.is_empty(), "the gloss list was found and parsed");
    parser.sort();
    shown.sort();
    assert_eq!(shown, parser, "the Agents panel glosses exactly the keys the parser reads");
}

/// …and the gloss of `role` names the jobs that EXIST. It named `summarizer`
/// for three increments after the role was deleted, which is worse than a
/// missing key: a person following it writes a line the parser now refuses.
#[test]
fn the_role_gloss_names_every_role_and_no_deleted_one() {
    let (_, said) = glossed()
        .into_iter()
        .find(|(k, _)| k == "role")
        .expect("`role` is glossed");
    for role in agent::ROLES {
        assert!(said.contains(role), "the gloss of `role` names `{role}`: {said}");
    }
    assert!(
        !KEY_HELP.contains("summarizer"),
        "`summarizer` is not a role any more; the panel must not offer it"
    );
}

File: packages/agent/src/frontmatter.js
/**
 * THE YAML SUBSET AN AGENT FILE IS WRITTEN IN — ours, and deliberately not a
 * YAML library.
 *
 * `Bun.YAML` exists and it is a BUILD-TIME api: `bun build --target=browser`
 * emits the call verbatim and the page has no `Bun`. A person may author an
 * agent in this browser, so the file that decides what that agent may call has
 * to be readable AT RUNTIME, by us, or authoring is a feature that works until
 * somebody uses it.
 *
 * The subset is every shape the shipped files use and nothing else: `key:
 * value`, a bare `key:` opening a block of `- item` lines, and the inline
 * `[a, b]` form. WHAT IT CANNOT READ IT REFUSES BY NAME. A key nothing reads
 * is a setting that looks applied — `engine: reakt` parsed clean for eighteen
 * rounds while selecting nothing — so an unknown key is a refusal, and so is a
 * value of a shape this cannot honour. Silence must never fail towards more
 * capability: a dropped `tools:` line leaves the list empty, and an empty list
 * means EVERY built-in.
 * @module
 */

import { STAGES } from '@harness/kernel'

/** The two engines. `react` is the tool loop; `base` is one reply with no tools, enforced by `toolbox.js` rather than described in prose. */
export const ENGINES = /** @type {const} */ (['react', 'base'])

/** The jobs the core used to hardcode as the string literals `main` and `summarizer`, so renaming a folder silently unhooked the machinery. A file DECLARES its job and the core looks the holder up. */
export const ROLES = /** @type {const} */ (['entry', 'critic'])

/** One file this build will not read, as a value: WHERE, WHICH KEY, and the sentence a person acts on. A value and not a throw because skipping a broken agent is correct — staying silent about it is what is not — so the roster projects this beside the agents that did load. @typedef {{path: string, key: string, message: string}} Refusal */

/** How one key's value is read, and the spec field it lands on. The file spells keys in snake_case; the spec is JavaScript. @typedef {{field: string, kind: 'text'|'closed'|'list'|'whole'|'number', legal?: readonly string[]}} Slot */

/** @type {Record<string, Slot>} */
const FIELDS = {
  name: { field: 'name', kind: 'text' },
  description: { field: 'description', kind: 'text' },
  model: { field: 'model', kind: 'text' },
  space: { field: 'space', kind: 'text' },
  engine: { field: 'engine', kind: 'closed', legal: ENGINES },
  role: { field: 'role', kind: 'closed', legal: ROLES },
  stages: { field: 'stages', kind: 'list' },
  tools: { field: 'tools', kind: 'list' },
  faculties: { field: 'faculties', kind: 'list' },
  compact_at: { field: 'compactAt', kind: 'whole' },
  keep_recent: { field: 'keepRecent', kind: 'whole' },
  max_rounds: { field: 'maxRounds', kind: 'whole' },
  passes: { field: 'passes', kind: 'whole' },
  temperature: { field: 'temperature', kind: 'number' },
}

/** @param {string} path @param {string} key @param {string} message @returns {{refusal: Refusal}} */
export function refuse(path, key, message) {
  return { refusal: { path, key, message } }
}

/**
 * The frontmatter block, read into the fields it names. The values are typed by
 * construction — a `whole` key holds a number here or nothing does — so the
 * caller merges them over the defaults without re-checking their shapes.
 * @param {string} path @param {string} block the text between the two `---` lines
 * @returns {{values: Record<string, unknown>} | {refusal: Refusal}}
 */
export function readFrontmatter(path, block) {
  /** @type {Record<string, unknown>} */
  const values = {}
  let open = ''
  for (const raw of block.split('\n')) {
    const line = raw.trim()
    if (line === '' || line.startsWith('#')) continue
    if (line.startsWith('- ')) {
      // A `- item` under no open list is dropped and NOT fed to `tools:`,
      // which is what the Rust's catch-all did: silence towards capability.
      if (open !== '') push(values, open, unquote(line.slice(2)))
      continue
    }
    const at = line.indexOf(':')
    // The one shape this reader cannot read at all, and the only line that used
    // to be dropped in silence — `exec` on its own, or a shell command a person
    // pasted a line early, would leave the file parsing clean while carrying an
    // instruction nothing here ever reads.
    if (at < 0) return refuse(path, '', `${path} holds the line "${line}", which is neither a "key: value" nor a "- item" under an open key, so nothing in this build reads it.`)
    const set = setField(path, values, line.slice(0, at).trim(), unquote(line.slice(at + 1)))
    if ('refusal' in set) return set
    open = set.open
  }
  return knownStages(path, values) ?? { values }
}

/** One `key: value`, onto the field it names. The return says which key's block list the lines below now belong to. @param {string} path @param {Record<string, unknown>} values @param {string} key @param {string} value @returns {{open: string} | {refusal: Refusal}} */
function setField(path, values, key, value) {
  const slot = Object.hasOwn(FIELDS, key) ? FIELDS[key] : undefined
  if (!slot) {
    return refuse(path, key, `${path} declares "${key}:", and no agent file key is called that — the keys are: ${Object.keys(FIELDS).join(', ')}.`)
  }
  // Last-wins is a silent choice made on the author's behalf. The `- item`
  // lines below an open key are exempt by construction: they go through
  // `push`, which appends, which is what a list under a key means.
  if (Object.hasOwn(values, slot.field)) {
    return refuse(path, key, `${path} declares "${key}:" twice, and only its author knows which was meant.`)
  }
  if (slot.kind === 'list') {
    const inline = value.startsWith('[') && value.endsWith(']')
    if (inline) values[slot.field] = split(value.slice(1, -1))
    else if (value === '') { values[slot.field] = []; return { open: slot.field } }
    else return refuse(path, key, `${path} holds "${key}: ${value}", and this build reads it as a list — write ${key}: [a, b], or a bare "${key}:" with "- name" lines under it.`)
    return { open: '' }
  }
  const read = scalar(path, key, value, slot)
  if ('refusal' in read) return read
  values[slot.field] = read.value
  return { open: '' }
}

/** @param {string} path @param {string} key @param {string} value @param {Slot} slot @returns {{value: unknown} | {refusal: Refusal}} */
function scalar(path, key, value, slot) {
  if (slot.kind === 'text') return { value }
  // A KEY WRITTEN WITH NOTHING AFTER IT IS A KEY NOBODY DECIDED. `whole` caught
  // it and `Number('')` did not, so `temperature:` alone ran an agent fully
  // deterministic at 0 — the default nobody chose, arrived at by arithmetic.
  // `role:` is the one blank that means something: a file holding no job.
  if (value === '' && key !== 'role') {
    return refuse(path, key, `${path} holds "${key}:" with nothing after it, and a value that parses to nothing would run as a default nobody chose.`)
  }
  if (slot.kind === 'closed') {
    if (value === '') return { value: '' }
    const legal = slot.legal ?? []
    return legal.includes(value)
      ? { value }
      : refuse(path, key, `${path} holds "${key}: ${value}", and the ${key} is one of: ${legal.join(', ')}.`)
  }
  const number = slot.kind === 'whole' ? whole(value) : Number(value)
  return Number.isFinite(number)
    ? { value: number }
    : refuse(path, key, `${path} holds "${key}: ${value}", and this build reads it as ${slot.kind === 'whole' ? 'a whole number' : 'a number'} — a value that parses to nothing would run as a default nobody chose.`)
}

/** A non-negative integer, or NaN. `parseInt` is not used: it reads "8 turns" as 8. @param {string} value @returns {number} */
function whole(value) {
  return /^\d+$/.test(value.trim()) ? Number(value) : Number.NaN
}

/** @param {Record<string, unknown>} values @param {string} field @param {string} item */
function push(values, field, item) {
  const held = values[field]
  if (Array.isArray(held) && item !== '') held.push(item)
}

/** The inside of an inline `[a, b]`. Empty items are dropped, so a trailing comma costs nothing; an empty LIST stays empty and means something. @param {string} inline @returns {string[]} */
function split(inline) {
  return inline.split(',').map(unquote).filter((s) => s !== '')
}

/** Quotes are a YAML nicety and never part of a value: `name: "shopper"` and `name: shopper` are one agent. @param {string} value @returns {string} */
export function unquote(value) {
  const v = value.trim()
  for (const q of ['"', "'"]) {
    if (v.length >= 2 && v.startsWith(q) && v.endsWith(q)) return v.slice(1, -1)
  }
  return v
}

/** Every stage name, from either list form, checked in one place — `engine`'s rule for the key that decides the whole loop. @param {string} path @param {Record<string, unknown>} values @returns {{refusal: Refusal} | null} */
function knownStages(path, values) {
  const named = Array.isArray(values['stages']) ? values['stages'] : []
  const bad = named.find((s) => !(/** @type {readonly unknown[]} */ (STAGES)).includes(s))
  return bad === undefined
    ? null
    : refuse(path, 'stages', `${path} names the stage "${bad}", and the stages are: ${STAGES.join(', ')}.`)
}
